### ISSUE TITLE
Cant delete file because its scheduled in the future

### ISSUE THREAD

**Issue by aresstokrat** [LibreTime/libretime#237]
LT dont let me remove file from library. 
I removed it from curent playing smart block instance and there is the schedule of this track in this future. Rechecked, 100% its not in the scheldule.
Tested on 2 tracks. Same problem.

**Comment by yfyf** [comment]
Can confirm, adding a track to a dynamic smart block and then scheduling it into the future will mark it as scheduled and unless you remove the actual scheduled show, you can't delete the track.

Would great to be able to override this constraint and or by default to allow to delete all tracks scheduled due to a smart block.

**Comment by yfyf** [comment]
I'd be willing to contribute if someone could guide me where exactly I should look at in the code to fix this.

**Comment by Robbt** [comment]
Ok, so this is probably a bug in terms of what happens when a smartblock is removed from a show. I don't know in particular where the code that marks a show scheduled or unscheduled is. Searching for that would be the first part of the process.

**Comment by Robbt** [comment]
This could be considered connected to #284

**Comment by Robbt** [comment]
People are still affected by this bug. Someone needs to dig into the code and find out why a track can be marked scheduled even when all instances are deleted. See https://discourse.libretime.org/t/please-help-could-not-delete-file-because-it-is-scheduled-in-the-future/540

**Comment by Robbt** [comment]
So it turns out there is a function             Application_Model_StoredFile::updatePastFilesIsScheduled();
that goes through and checks files to make sure those that are no longer scheduled are marked unscheduled.

It is triggered when a show is cancelled and when $scheduler->removeItems is ran so it should happen every time items are removed from a show. 

So there is a still a bug lurking here where shows are cancelled and this isn't ran but according to the forum post creating a new show and cancelling it released the lock on all of the future files.

TLDR, workaround - create a new show and cancel it

I'm going to close this until someone can come up with a set of steps to recreate this scenario.

**Comment by Jeroen70** [comment]
Perhaps my comment is totally unrelated, but here we go:

i installed Libre Time (wonderful), uploaded tracks, scheduled shows and so on, so nothing new, apart from the pretty radom playout issue, however thats not the main thing here.

In other to see if some tracks are corrupt, i thought i could delete most of them. I opted to use the delete function (settings --> general --> delete all tracks). This didnt work. Ok, so i tried deleting them manually. No succes as some tracks were scheduled for future shows. Ok i guess, so the next step was to delete all shows, smart blocks and playlists. No problem, but when i tried deleting the tracks manually again, the same info box pops up saying some tracks were scheduled for future use. Kinda weird, but ok. So i check the ftp and delete the tracks i dont need through ftp.

Now here its gets funny. Once deleted through ftp, one could imagine the files will not longer be present. Well, guess agian.......the are still visible.

Long story short: ........ I check which files are scheduled for future use by displaying the column in the tracks overview. All tracks that are not listed as scheduled could be deleted manually. 

Funny enough.......you cant seem to delete the tracks onse they were once scheduled for future shows. Deleting the shows should actually 'unschedule' the tracks of those shows, however this doesntseem to happen.

Doed anyone know how to remove all music by using the command screen?

**Comment by robz0** [comment]
Hi Sorry to revive this very old thread but it is neither stale or resolved. 

In Status - Libretime Version says - f809c3a8f.  Though not sure what that means. I konw it has been updated in the last couple of months. 

Looking at RobbT's suggestions of creating a new show then deleting. This did not work for me. Just created a show. Scheduled an old 'flagged as scheduled' track in - just in case. Then cleared show and deleted show. But nothing happened. If this was to trigger the:  function Application_Model_StoredFile::updatePastFilesIsScheduled();
I don't think it did. 

Maybe suggest a couple of things;
1 - force delete. Given this bug will mean we can't delete tracks, then the storage fills up and new ones cannot be added. I would rather risk some dead air than have no space. 

2- find scheduled tracks - though the tracks are not really scheduled so this may not be any good. 

3 - some patch that would allow front-end management and (forced) deletion -as we don't have back end server access for getting into the database etc.

Any further updates and guidance, very welcome. 
Thanks
Rob

**Comment by robz0** [comment]
or 4? - a button to force this function
function Application_Model_StoredFile::updatePastFilesIsScheduled()

...is this is the problem issue.  
Thanks for identifying it, if so. 
Rob

**Comment by robz0** [comment]
Hi 
I wondered if any more knowledgeable members of the LibreTeam had time to think about solutions for this one. 
Or if any related patches might be applied to upcoming releases of LT.

some of the ideas that stick to me are:

- Force Delete (regardless of scheduled in the future) - given that we would prefer to delete  and create a gap, rather than run out of disk space altogether.

- Find Scheduled track in schedule - although, this may be a wild goose chase if this is really a bug and the track is not actually scheduled

- (Front-end) Options to restart services and dependencies - as there was [some discussion](https://discourse.libretime.org/t/please-help-could-not-delete-file-because-it-is-scheduled-in-the-future/540/28) about this being a problem with such things being in need of resetting. 

In terms of **Triage Priority**- the consequence is that space will run out when tracks cannot be deleted. And server space costs go up.  So for us the priority is HIGH. Also it does not seem to be isolated to our instance and may be a wider problem for other users.

Also, want to stress that it would be good (for us) to have frontend control of such things like Resetting services, as we have no backend access. 

Much Thanks.
Rob0

**Comment by robz0** [comment]
> I think a force option would probably be the easiest way forward.
> 
> Restarting services via the UI is tricky, as it requires privilege escalation for installations where the systemd units are not run as the web-server user. The default on Debian and Ubuntu is that they are, so there it could be possible.

Thanks for your attention to this Kyle, 

From a lay person's point of view, I think you are right. Would also definitely be easier from a User perspective.

Hopefully such an option would be relatively easy to apply to LT, without bugging up whatever underlying services are managing 'scheduled' content.

Rob
